app/controllers/redirect/redirect_controller.rb

```diff
@@ -10,14 +10,35 @@ def location_params
       params.permit(*PERMITTED_LOCATION_PARAMS).to_h.symbolize_keys
     end
 
+    def partner_query_params
+      @partner_query_params ||= begin
+        {
+          partner: current_sp&.agency&.name,
+          partner_div: current_sp&.integration&.name,
+        }.compact
+      end
+    end
+
+    def add_partner_query_params(url)
+      uri = Addressable::URI.parse(url)
+
+      if partner_query_params.any?
+        uri.query_values = (uri.query_values || {}).merge(partner_query_params)
+      end
+
+      uri.to_s
+    end
```

Comment:
I would do this and then probably just inline this method as well
```suggestion
    def add_partner_query_params(url)
      UriService.add_params(url, partner_query_params)
    end
```

Reply:
@zachmargolis I've updated `UriService` so that it will not add an empty question mark when given empty params. This would previously result in URLs like `https://www.login.gov/help/article/?#heading` and `https://www.login.gov/help/?`.

`Addressable::URI` seems to offer some encoding advantages over plain `URI()`. However, I have removed references`Addressable::URI` and used `UriService` instead.

Reply:
awesome! thanks for cleaning that up